### ISSUE TITLE
Cython annotated `distributed/scheduler.py`

### ISSUE THREAD

**Issue by jakirkham** [dask/distributed#4484]
Thought people curious about the Scheduler Cythonization effort might be interested in getting a glimpse of what the code generated by Cython looks like. Using commit ( https://github.com/dask/distributed/commit/a0d60e7b2ad16304128c7989a111539a4d80ed38 ), generated [the annotated HTML file]( https://gistcdn.githack.com/jakirkham/264d50e33d74c1363af016151b1ace35/raw/4e230bbe05a662d1b2ed9537689c7754c67bb68b/scheduler.html ). As we do use Python objects in a lot of place, there is always some code generated around these to handle reference counting, confirming the object is not `None` before accessing, raising any errors that come up, etc. That said, a lot of these operations are pretty quick. So the level of highlighting may not be fully indicative of how much time a piece of code actually takes to run in practice. IOW view this with a grain of salt. In any event, take a look. Feel free to share any thoughts you may have 🙂

**Comment by jakirkham** [comment]
cc @mrocklin @quasiben

**Comment by martindurant** [comment]
@jakirkham 
Since I've had quite a bit of cython practice with fastparquet lately, I'd be happy to help adding types. Do you have a list of outstanding areas being compiled but not annotated?

(NB this is the first issue that I found when searching - things probably have moved on and this one should b closed)

**Comment by jakirkham** [comment]
Probably wrapping up the `update_graph` work ( https://github.com/dask/distributed/pull/4418 ) would be a good place to start.

Could also use your help reviewing PR ( https://github.com/dask/distributed/pull/4650 ). Planning to rework that next week so can ping once it is ready (shouldn't take too long).

There have also been some finicky things historically where typing some attributes hasn't worked well ( https://github.com/dask/distributed/pull/4336 ) ( https://github.com/dask/distributed/pull/4296 ). Or other things currently typed as `object` ([`_run_spec` for example]( https://github.com/dask/distributed/blob/09e843662b55f1bbb4ded941b8fce1398110230e/distributed/scheduler.py#L1283 )). Some of these may involve figuring out how we intend to use those objects and then typing them appropriately. Some may benefit from narrowing their intended usage a bit.

One of things @madsbk and @quasiben and I have discussed is constraining the kinds of messages that are passed from Workers to the Scheduler and vice versa. This probably involves some knowledge of the kinds of messages that are used today. The benefit here is this would speed up message construction, communication, and usage on the Scheduler. These would be new classes that would be need to built (maybe with an eye towards using [msgspec]( https://github.com/jcrist/msgspec )). This is also relatively green field atm.

There are various other functions that haven't been on the critical path (adding & removing workers, dashboard queries, etc.). That haven't been annotated as much, but are probably lower priority (though if you have a use case, they may make sense to look into).

Somewhat unrelated, but one thing we have discussed is it might be worthwhile to refactor out [`SchedulerState`]( https://github.com/dask/distributed/blob/09e843662b55f1bbb4ded941b8fce1398110230e/distributed/scheduler.py#L1714 ) so that the [`Scheduler`]( https://github.com/dask/distributed/blob/09e843662b55f1bbb4ded941b8fce1398110230e/distributed/scheduler.py#L3198 ) doesn't inherit from it and instead contains it as an object. This might allow the Cythonization of `SchedulerState` and related objects to move into its own module.

cc @mrocklin @fjetter @jrbourbeau (in case any of you have thoughts here)

**Comment by jakirkham** [comment]
> NB this is the first issue that I found when searching - things probably have moved on and this one should b closed

Also yes this could be closed. Was trying to share where things stood earlier on. Though am also fine repurposing this for general discussion

**Comment by martindurant** [comment]
Thanks - I'll have a dig around next week.

**Comment by jakirkham** [comment]
Sounds good. Thanks Martin!

(also added one more note about test failures we've identified when Cythonizing)
